### PR TITLE
researcher: Prefer CPID with active beacon for primary CPID

### PR DESCRIPTION
**Pull request by scribblemaniac**
In a situation where multiple CPIDs are detected, Researcher should prefer CPIDs with an active beacon over CPIDs without an active beacon when determining its primary CPID. This change will select the first CPID with an active beacon in the project list, or if there is no such CPID, it will select the last CPID in the project list as before.

The motivation for this change came from my recent experience with finally going solo. I sent my beacon and got everything working very easily. However when I went to add another project, and when to check the projects page of the researcher wizard, it took me back to the beginning of the wizard. I quickly figured out that it had selected the CPID of the new project and just waited for BOINC to resolve the split-CPID before rechecking, but I could see this being a very confusing behavior, especially for someone new to Gridcoin. In this case, there is an obvious choice for the primary CPID. I'm not sure if this is the best way to implement it, but it seems like a reasonable place for the check to me, and in testing it works as I expected it to.

### CODE REVIEW

**Comment by scribblemaniac** [comment]
It occurs to me as soon as I have submitted this that the code could be a bit neater and more efficient if TryProjectCpid returned a bool for success / failure. Then you could do something like this:
```
if (TryProjectCpid(mining_id, project_pair.second)) {
    LOCK(cs_main);
    if (GetBeaconRegistry().ContainsActive(mining_id.TryCpid())) {
        break;
    }
}
```
Whatever you prefer.

File: src/gridcoin/researcher.cpp [hunk 1]
```diff
@@ -1195,6 +1195,14 @@ void Researcher::Reload(MiningProjectMap projects, GRC::BeaconError beacon_error
 
     if (mining_id.Which() != MiningId::Kind::CPID) {
         for (const auto& project_pair : projects) {
+            // Stop searching if the current mining_id already has an active beacon
+            if (const CpidOption cpid = mining_id.TryCpid()) {
+                LOCK(cs_main);
```
Comment (jamescowens) [inline src/gridcoin/researcher.cpp]: I don't think you need a lock on cs_main here. cs_main is always supposed to be taken before calling any of these functions, and putting in an extra lock might seem harmless, but in fact could cause a deadlock if another lock from another thread gets taken in between.

**Comment by jamescowens** [comment]
This PR is a nice enhancement BTW.

File: src/gridcoin/researcher.cpp (hunk 1 above)
Comment (scribblemaniac) [reply src/gridcoin/researcher.cpp]: Right, I see that's documented in the .h file. I have removed the lock now.

**Review by jamescowens — APPROVED** [review]
I think this is good. It doesn't affect the possible split notification, which is downstream, and I agree that if a beacon has already been advertised and is active, that CPID should be preferred. utACK.

**Pull request merged**